### PR TITLE
feat: Add configurable max limit for TEI query [DHIS2-10679]

### PR DESCRIPTION
**Pull request by zubaira**

### CODE REVIEW

File: dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/trackedentityinstance/TrackedEntityInstanceQueryLimitTest.java [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+import org.hisp.dhis.user.UserService;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+
+import com.google.common.collect.Sets;
+
+/**
+ * @author Zubair Asghar
+ */
+class TrackedEntityInstanceQueryLimitTest extends TransactionalIntegrationTest
```
Comment (teleivo) [inline dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/trackedentityinstance/TrackedEntityInstanceQueryLimitTest.java:61]: @enricocolasante what is the preferred one now, SingleSetupIntegrationTestBase? I think this one is slower at least when we would add more tests into this test class, right?

File: dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/trackedentityinstance/TrackedEntityInstanceQueryLimitTest.java [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+        params.setProgram( program );
+        params.setOrganisationUnits( Sets.newHashSet( orgUnitA ) );
+        params.setOrganisationUnitMode( OrganisationUnitSelectionMode.ALL );
+        params.setUser( user );
+        params.setSkipPaging( true );
+
+        List<Long> teis = trackedEntityInstanceService.getTrackedEntityInstanceIds( params,
+            false, false );
+
+        assertThat( teis, hasSize( is( 3 ) ) );
```
Comment (teleivo) [inline dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/trackedentityinstance/TrackedEntityInstanceQueryLimitTest.java:169]: Please remove the assertion on the size since the later hamcrest contains assertion will fail if there is an element which is not in your call to `contains`.

For example

```java
        assertThat(Arrays.asList("foo", "bar", "free"), contains("foo", "bar"));
```

leads to

```
java.lang.AssertionError: 
Expected: iterable containing ["foo", "bar"]
     but: not matched: "free"
```

which gives us the necessary context to fix an issue with our test/implementation. If the test were to become flaky/only fail on Github.

With the current assertions the integration test would fail on unequal size of expected/actual collections with no more context.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/hibernate/HibernateTrackedEntityInstanceStore.java [hunk 3]
```diff
@@ -145,22 +147,26 @@
...
-        checkNotNull( statementBuilder );
-        checkNotNull( organisationUnitStore );
+        checkNotNull( instanceStoreServicesSupplier );
 
-        this.statementBuilder = statementBuilder;
-        this.organisationUnitStore = organisationUnitStore;
+        this.instanceStoreServicesSupplier = instanceStoreServicesSupplier;
+    }
+
+    @PostConstruct
```
Comment (teleivo) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/hibernate/HibernateTrackedEntityInstanceStore.java:166]: `@PostConstruct` should not be used IMHO 😄 I think this leads to bad design. In order to for example test such a class in a unit test, you need to call init after having constructed an object of the class. If you forget to do that in a test your test diverges from what happens in production.

It looks to me as if HibernateTrackedEntityInstanceStore and the new TrackedEntityInstanceStoreServicesSupplier both depend on the StatementBuilder. If so let them both depend on it and Spring inject the StatementBuilder. If I did not make any mistake in reading the code we would then not need the PostConstruct 😅

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/TrackedEntityInstanceStoreServicesSupplier.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+import org.hisp.dhis.organisationunit.OrganisationUnitStore;
+import org.springframework.stereotype.Component;
+
+/**
+ * @author rajazubair
+ */
+@Component
+@RequiredArgsConstructor
+@Getter
+public class TrackedEntityInstanceStoreServicesSupplier
```
Comment (teleivo) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/TrackedEntityInstanceStoreServicesSupplier.java:45]: Do we even need this class? If I understand the code correctly you would only need to add the DhisConfigurationProvider to the HibernateTrackedEntityInstanceStore 😅

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/hibernate/HibernateTrackedEntityInstanceStore.java [hunk 5]
```diff
@@ -1360,10 +1371,21 @@ private String getFromSubQueryLimitAndOffset( TrackedEntityInstanceQueryParams p
     {
         StringBuilder limitOffset = new StringBuilder();
         int limit = params.getMaxTeiLimit();
+        int teiQueryLimit = Integer
+            .parseInt( instanceStoreServicesSupplier.getConfigurationProvider()
```
Comment (teleivo) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/hibernate/HibernateTrackedEntityInstanceStore.java:1375]: It seems to me that we do not seem to validate the config and fail DHIS2 on startup if the config is invalid. This is unfortunate since a user can put in a String like `foo` into this config and it will only manifest when the first user makes a request that calls this store. If we do have the ability to validate the config and fail DHIS2 startup then lets do it 😄 If not, can we at least parse this only once maybe in the constructor and store it in a field?

**Review by teleivo — CHANGES_REQUESTED** [review]
I have some suggestions 😄

File: dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/trackedentityinstance/TrackedEntityInstanceQueryLimitTest.java (hunk 1 above)
Comment (enricocolasante) [reply dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/trackedentityinstance/TrackedEntityInstanceQueryLimitTest.java:61]: Agreed.
With one test there is no difference but if we will had more tests the other one is better

**Review by enricocolasante — COMMENTED** [review]
Nothing to add to what @teleivo already commented on

File: dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/trackedentityinstance/TrackedEntityInstanceQueryLimitTest.java [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        TrackedEntityInstanceQueryParams params = new TrackedEntityInstanceQueryParams();
+        params.setProgram( program );
+        params.setOrganisationUnits(Sets.newHashSet( orgUnitA ) );
+        params.setOrganisationUnitMode( OrganisationUnitSelectionMode.ALL );
+        params.setUser( user );
+        params.setSkipPaging(true);
+
+        List<Long> teis = trackedEntityInstanceService.getTrackedEntityInstanceIds( params, false, false );
+
+        assertThat( teis, Matchers.hasSize( Matchers.is( 3) ) );
```
[the remark this answers is not in the document]
Comment (zubaira) [reply dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/trackedentityinstance/TrackedEntityInstanceQueryLimitTest.java]: yeah agreed. that is was i am looking into how we can use different values for ConfigurationKey for different tests.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/TrackedEntityInstanceStoreServicesSupplier.java (hunk 4 above)
Comment (zubaira) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/TrackedEntityInstanceStoreServicesSupplier.java:45]: yes we do need this. Otherwise just injecting DhisConfigurationProvider in constructure will introduce some sonarqube  code smell. This was one of the ways to avoid that without having to introduce much bigger change.

Comment (zubaira) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/TrackedEntityInstanceStoreServicesSupplier.java:45]: there is a limit set on the number of parameters we can have in the constructor.

File: dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/trackedentityinstance/TrackedEntityInstanceQueryLimitTest.java (hunk 2 above)
Comment (stian-sandvold) [reply dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/trackedentityinstance/TrackedEntityInstanceQueryLimitTest.java:169]: Since we are testing that the limit is 3, would it make sense to assert the size is not >3?  @teleivo 

I would say potentially in this case the contains is more flaky, incase the default ordering is changing (Which we might actually do); Or @zubaira you can specify the order in the test to make sure its deterministic throughout changes like that

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/hibernate/HibernateTrackedEntityInstanceStore.java [hunk 7]
```diff
@@ -145,22 +147,28 @@
     // Dependencies
     // -------------------------------------------------------------------------
 
-    private final OrganisationUnitStore organisationUnitStore;
```
Comment (teleivo) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/hibernate/HibernateTrackedEntityInstanceStore.java:148]: Why are you removing the `final` on the dependencies? I think we should keep it

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/hibernate/HibernateTrackedEntityInstanceStore.java [hunk 8]
```diff
@@ -1360,10 +1373,19 @@ private String getFromSubQueryLimitAndOffset( TrackedEntityInstanceQueryParams p
     {
         StringBuilder limitOffset = new StringBuilder();
         int limit = params.getMaxTeiLimit();
+        int teiQueryLimit = systemSettingManager.getIntSetting( SettingKey.TRACKED_ENTITY_MAX_LIMIT );
 
         if ( limit == 0 && !params.isPaging() )
         {
-            return "";
+
+            return Optional.of( teiQueryLimit ).filter( a -> teiQueryLimit == 0 ).map( a -> "" )
```
Comment (teleivo) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/hibernate/HibernateTrackedEntityInstanceStore.java:1381]: I think we should simplify this into

```java
        if (teiQueryLimit > 0) {
            limitOffset
                    .append("limit")
                    .append(" ")
                    .append(teiQueryLimit)
                    .append(" ");
        }

        return limitOffset.toString();
```

I also assume a limit `teiQueryLimit < 0` does not make sense hence the `teiQueryLimit > 0` check.

File: dhis-2/dhis-support/dhis-support-external/src/main/java/org/hisp/dhis/external/conf/ConfigurationKey.java [hunk 9]
```diff
@@ -702,7 +702,6 @@
      * other session when a new session is started.
      */
     MAX_SESSIONS_PER_USER( "max.sessions.per_user", "10", false ),
-
```
Comment (teleivo) [inline dhis-2/dhis-support/dhis-support-external/src/main/java/org/hisp/dhis/external/conf/ConfigurationKey.java:705]: Can you please remove unrelated changes. Might be accidental 😄

File: dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/trackedentityinstance/TrackedEntityInstanceQueryLimitTest.java (hunk 1 above)
Comment (teleivo) [reply dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/trackedentityinstance/TrackedEntityInstanceQueryLimitTest.java:62]: @zubaira now that you added more tests can you please change to SingleSetupIntegrationTestBase. I assume your tests are fine if `setUpTest` is only run once.

File: dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/trackedentityinstance/TrackedEntityInstanceQueryLimitTest.java [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+        userService.addUser( user );
+    }
+
+    @Test
+    void testConfiguredMaxTeiLimit()
+    {
+        systemSettingManager.saveSystemSetting( SettingKey.TRACKED_ENTITY_MAX_LIMIT, 3 );
+        TrackedEntityInstanceQueryParams params = new TrackedEntityInstanceQueryParams();
+        params.setProgram( program );
+        params.setOrganisationUnits( Sets.newHashSet( orgUnitA ) );
```
Comment (teleivo) [inline dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/trackedentityinstance/TrackedEntityInstanceQueryLimitTest.java:166]: Please prefer Java's `Set.of` for something like that 😄 (mind https://github.com/dhis2/wow-backend/blob/master/guides/java11_pitfalls.md)

File: dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/trackedentityinstance/TrackedEntityInstanceQueryLimitTest.java [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+        params.setOrganisationUnitMode( OrganisationUnitSelectionMode.ALL );
+        params.setUser( user );
+        params.setSkipPaging( true );
+
+        List<Long> teis = trackedEntityInstanceService.getTrackedEntityInstanceIds( params,
+            false, false );
+
+        assertNotNull( teis );
+        assertEquals( 4, teis.size() );
```
Comment (teleivo) [inline dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/trackedentityinstance/TrackedEntityInstanceQueryLimitTest.java:192]: assertContainsOnly already checks the size using assertEquals 😄 So please remove assertEquals.

**Review by teleivo — CHANGES_REQUESTED** [review]

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/hibernate/HibernateTrackedEntityInstanceStore.java (hunk 8 above)
Comment (teleivo) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/hibernate/HibernateTrackedEntityInstanceStore.java:1381]: @zubaira can you please address this 😄

**Review by teleivo — CHANGES_REQUESTED** [review]
@zubaira you said in the standup that the PR conversations have been resolved, but there are still a bunch of open conversations. Can you please have a look 😄

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/hibernate/HibernateTrackedEntityInstanceStore.java (hunk 8 above)
Comment (zubaira) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/hibernate/HibernateTrackedEntityInstanceStore.java]: resolved

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/hibernate/HibernateTrackedEntityInstanceStore.java [hunk 12]
```diff
@@ -1360,10 +1373,18 @@ private String getFromSubQueryLimitAndOffset( TrackedEntityInstanceQueryParams p
     {
         StringBuilder limitOffset = new StringBuilder();
         int limit = params.getMaxTeiLimit();
+        int teiQueryLimit = systemSettingManager.getIntSetting( SettingKey.TRACKED_ENTITY_MAX_LIMIT );
 
         if ( limit == 0 && !params.isPaging() )
         {
-            return "";
+
+            return Optional.of( teiQueryLimit ).filter( a -> teiQueryLimit > 0 ).map( a -> limitOffset
```
Comment (teleivo) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/hibernate/HibernateTrackedEntityInstanceStore.java:1381]: This one https://github.com/dhis2/dhis2-core/pull/11321/files/3f2d605b48a848b713fe96f2d85c88e2cb5049bc#r955991464
is not resolved.

Pasting it here in case it got lost for you 😄 

I think we should simplify this into

```java
        if (teiQueryLimit > 0) {
            limitOffset
                    .append("limit")
                    .append(" ")
                    .append(teiQueryLimit)
                    .append(" ");
        }

        return limitOffset.toString();
```

I also assume a limit `teiQueryLimit < 0` does not make sense hence the `teiQueryLimit > 0` check.

**Review by teleivo — CHANGES_REQUESTED** [review]
one issue is still in 😄

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/hibernate/HibernateTrackedEntityInstanceStore.java (hunk 12 above)
Comment (zubaira) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/hibernate/HibernateTrackedEntityInstanceStore.java:1381]: existing code is doing the same thing with the only difference of using `Optional` instead of `If` statement

Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/hibernate/HibernateTrackedEntityInstanceStore.java:1381]: I really think that the if version is much more readable.
@zubaira You are right that they are doing the same thing but the use of Optional here is just confusing.
Optional should be used for variables that maybe are present and maybe they are not (null), in this case `teiQueryLimit ` can never be null, especially because it is an `int`.
What you want here is to do something if `teiQueryLimit` is greater than 0 and do nothing otheerwise.
An if statement here is much more human readable than the use of Optional

**Review by teleivo — APPROVED** [review]
Thank you for the change @zubaira 😄

**Review by lucaCambi77 — APPROVED** [review]

**Comment by teleivo** [comment]
@stian-sandvold the smells are actually one smell that we need to tackle. I think we can merge this PR, but we should definitely schedule this work.
